### PR TITLE
Capture tab key press in CLI and forward command to fc with appended …

### LINKED ISSUE #993: [Feature Request] Adding tab-based autocomplete to CLI.
I think it would be super handy, especially with the number of long commands that can be typed/

### PR DESCRIPTION
**Pull request by ademuk**
…tab character for auto completion

Fixes #993

- Triggers and outputs [Betaflight auto complete functionality](https://github.com/betaflight/betaflight/blob/master/src/main/interface/cli.c#L3913)
- Keeps track of the firmware's `cliBuffer` so only diffs are sent. Including backspace escape characters if prompt characters are deleted after having previously been sent to firmware (via tab key).
- Grabs the last command, if multiple have been pasted into the cli.
- Tab key detection is on `keydown`. By the time `keypress`/`keyup` have happened, `textarea` has lost focus.
- No firmware change required

### CODE REVIEW

**Comment by mikeller** [comment]
Nice one!

I think this should be improved a little, probably by not sending the `\n` after the tab. If the prefix is not unique when tab is hit, this currently causes an error (for example after typing `a<tab>`:

```
# djrange	aux	
# a
Unknown command, try 'help'
# 
```

Probably just send tab, and read back the output. If the prefix is unique, the firmware will complete the command on the same line, so grab this, and append it to the input, and wait for the user to hit enter. This makes it possible to use tab for commands that expect parameters, plus it fixes the big issue of commands being expanded and executed without giving the user a chance to review. (i.e. typing `de<tab>` with what we have now will execute `defaults`, and wipe the configuration).

**Comment by ademuk** [comment]
Hey @mikeller, thanks for your quick reply.

My change is actually only useful for completing `get` commands. `cliGet` get's invoked, and all matching commands are displayed in the cli output.

I like the idea of making it more generic, making use of the the existing auto completion running on fc. I have a few queries about your suggestions:
* In the case that the fc finds a unique match and sends it back completed (with trailing space), do you have any suggestions on how client would differentiate this response from normal output, so that it knows to rewrite the prompt's `textarea`?
* In the case that there are multiple matches, the fc returns escape sequence `\r\033[K`, presumably in attempt to move the cursor to the start of the line and clear the line, before returning matching tab separated commands. The client doesn't know how to interpret this escape seq, but instead fails to display the output at all. I have a fix which interprets this escape seq on the client and fixes the display of matching commands. Is this the right path to take?
* With commands that take params (i.e. `get <var>`). I can't see code on the fc that is going to complete those variable names for us, am I missing something?

**Comment by mikeller** [comment]
I think since we already know if we've sent a request for tab completion, we only have to distinguish 3 cases, and then act appropriately:
- no match => do nothing;
- exactly one match => complete the line in the editor, wait for input;
- more than one matches => probably again nothing, the choices are already displayed in the output area.

Tab completion only works for commands. For parameter names, there is no tab completion, but a `get <prefix>` will return you settings for all parameters that match the prefix.

**Comment by ademuk** [comment]
@mikeller Thanks for the tips.

I've updated my PR to cover your 3 points of functionality. It' working great now.

It does however require [the following changes to the fc firmware](https://github.com/betaflight/betaflight/compare/master...ademuk:cli-tab-auto-complete):

1. Return partial auto complete output (i.e. `seri` > `serial`) before outputting unambiguous matched (i.e. `serial	serialpassthrough`)
2. Remove `\r\033[K` escape sequence which appears not to be working correctly.
3. Set `bufferIndex = 0` after auto-completion is returned. Fixes subsequent commands/tab key presses.

File: src/js/tabs/cli.js [hunk 1]
```diff
@@ -152,19 +164,23 @@ TABS.cli.read = function (readInfo) {
...
         if (CONFIGURATOR.cliValid) {
-            if (data[i] == 27 || this.sequenceElements > 0) { // ESC + other
+            const escapeSequenceCode = 27;
+            if (data[i] == escapeSequenceCode || this.sequenceElements > 0) { // ESC + other
                 this.sequenceElements++;
 
                 // delete previous space
                 if (this.sequenceElements == 1) {
-                    text = text.substring(0, text.length -1);
+                    cliOutput = text.substring(0, text.length - 1);
```
Comment (mikeller) [inline src/js/tabs/cli.js]: `text` here is causing ReferenceError.

Comment (ademuk) [reply src/js/tabs/cli.js]: Good spot, I’ll fix when I have laptop access.

I restored this last minute. Admittedly couldn’t see what it’s purpose was, do you know?

Also, is there any appetite for automated tests for the configurator? I’d be happy to add some for this file.

Comment (mikeller) [reply src/js/tabs/cli.js]: I don't know what it exactly does, no.

And yes, tests are very welcome. You can add the test runner to `package.json`.

**Comment by ademuk** [comment]
@mikeller `ReferenceError` fixed. 5 unit tests added. I've used Karma for minimal config browser testing. Let me know what you think.  

What do you think about the need for firmware changes, is that going to work?

Does the firmware cli have any other clients, i.e. Android?

**Comment by mikeller** [comment]
This looks very good now, good work.

One little snag that it still has: When you start typing a command, and hit tab at a point where the command is ambiguos, you get the selection in the output pane (slightly distorted but still useful), but the start of the command that you typed is also removed from the input field. If you then continue typing until the command is unambiguos and hit tab, the command that is selected is shown in the input field, but the letters you typed before hitting tab for the first time are missing. The following screenshot is after typing `a<tab>d<tab>`:

image

Do you think this can be fixed?

**Comment by ademuk** [comment]
@mikeller Are you running [my firmware changes](https://github.com/betaflight/betaflight-configurator/pull/995#issuecomment-378094004)? If so, which OS, Mac/Win?

**Comment by ademuk** [comment]
@mikeller Ok, I'll take another look.

Just out of curiosity, how is the cli accessible from a terminal?

**Comment by McGiverGim** [comment]
If I'm not wrong, I think remember you must send the `#` symbol to the COM port where is connected the FC, and the FC will enter in CLI mode and wait for commands.

**Comment by mikeller** [comment]
@ademuk: Yes, as @McGiverGim says, the 'enter CLI' command is `#` (conveniently). Send this on a serial connection, and the firmware drops into CLI mode.

**Comment by ademuk** [comment]
Hi @mikeller,

I’ve got some spare time coming up this week to work on this. My idea is to keep hold of a command buffer in memory, client side. This should allow us to build up the auto completed commands upon tab press, and only process it upon carriage return.

**Comment by ademuk** [comment]
@mikeller, I’ve got something working for this. Expect a pr in the next few days. I plan to squash existing commits.

**Comment by mikeller** [comment]
Thanks for letting me know, @ademuk.

**Comment by ademuk** [comment]
@mikeller Squashed, pushed and ready for review. I've updated the PR description too.

I've done some manual testing, so far I haven't been able to break it..

**Comment by mikeller** [comment]
@ademuk: So does this now work with the unmodified firmware, or does this need the firmware changes proposed in https://github.com/betaflight/betaflight-configurator/pull/995#issuecomment-378094004 ?

**Comment by ademuk** [comment]
No firmware change required. Since it’s now respecting the cliBuffer to render the prompt.

**Comment by mikeller** [comment]
:+1:

**Comment by McGiverGim** [comment]
Great feature!! But... how does this work? I'm trying to test it without too much luck, it does strange things.

If I write `gps` and press tab, the `gpspassthrough` command appears, but if I delete it and write `res` and press tab, it adds resources to the previous command in the log. 

It does not serve for parameters? something like `set gps` and tab does not work as expected.

I'm very curious about that :)

**Comment by ademuk** [comment]
Thanks for looking @McGiverGim!

So the first scenario is valid. `res` ends up being auto-completed to `resource` in your prompt right? This auto-complete data also ends up in the output. Do you think this is confusing to the user? It has no consequence to cli behaviour.

Secondly, there is no param support as the firmware does not currently support it. I'd love to see it for the `get` command. Maybe something for the future.

**Comment by McGiverGim** [comment]
I had several problems:
- One, I don't know if it is easily reproducible. I used it, disconnected, connect, and when I went again to the CLI and when trying to use it it appear the command that I typed before disconnect.
- The second problem is that it is concatenating things in the output, that is strange to me. Maybe if it does in different lines it will be easier for the user. If I start with `gps`+tab but I decide to not do that, and delete and type `res`+tab, in the output there are one line with two commands, and I think is better two lines.

All this without having played too much with it, maybe I'm not using it as expected.

**Comment by ademuk** [comment]
That’s good feedback thanks. I think I can resolve the first issue, I don’t think the buffer should be outputted. The second is a trivial one. TABS.cli.cliBuffer needs clearing upon exit/cleanup. Leave it with me.

**Comment by ademuk** [comment]
@McGiverGim Try now. I've added a couple of tests to cover those scenarios too.

**Comment by mikeller** [comment]
@ademuk:

That was an entirely accidental spot. I was done testing this feature, and wanted to test something else, and noticed that the board didn't reset when leaving the CLI tab...

Re the leading characters: Good point. I am on linux. The reality is that MacOsX and linux are much alike, and Windows is the odd one out, and this proves true here too. The following change will make the preview look fine on linux:

```
diff --git a/src/js/tabs/cli.js b/src/js/tabs/cli.js
index 2a724953..608f12e9 100644
--- a/src/js/tabs/cli.js
+++ b/src/js/tabs/cli.js
@@ -254,13 +254,13 @@ TABS.cli.read = function (readInfo) {
 
         switch (data[i]) {
             case lineFeedCode:
-                if (GUI.operating_system != "MacOS") {
+                if (GUI.operating_system === "Windows") {
                     writeLineToOutput(this.cliBuffer);
                     this.cliBuffer = "";
                 }
                 break;
             case carriageReturnCode:
-                if (GUI.operating_system == "MacOS") {
+                if (GUI.operating_system !== "Windows") {
                     writeLineToOutput(this.cliBuffer);
                     this.cliBuffer = "";
                 }
```

**Comment by ademuk** [comment]
I wonder whether we should be parsing for CRLF pairs (\r\n), and render `<br>` for them. As that's what the firmware seems to output.

And when we see the anomaly `\r\033[K` (ambiguous results) - which is "carriage return & erase until end of line", render a `<br>` regardless of operating system.

@mikeller  I've pushed your suggested change though. The leading char issue would only exist on Windows now.

**Comment by mikeller** [comment]
@ademuk: Thanks for making the changes. I am mergeing this for now to make sure it does not get stale, as we are preparing to ramp up for a release. But feel free to investigate a fix for the Windows version, and open a new pull request if you've got something.

**Pull request merged**